### PR TITLE
fix: preserve leading zero bytes in base58btc decode

### LINKED ISSUE #155: _b58decode drops leading zero bytes (base58btc spec violation)
## Summary

`_b58decode` in `src/didkey.py` (line 73-80) does not preserve leading `0x00` bytes. In base58btc, leading `'1'` characters encode leading zero bytes, but the current implementation uses pure integer arithmetic which loses them `int.from_bytes` followed by `.to_bytes` strips any leading zeros from the result.

## The bug

```python
def _b58decode(raw: str) -> bytes:
    n = 0
    for ch in raw:
        digit = _B58_INDEX.get(ch)
        if digit is None:
            raise DidError(f"bad did:key: {ch!r} is not base58btc")
        n = n * 58 + digit
    return n.to_bytes((n.bit_length() + 7) // 8, "big") if n else b""
```

A payload whose raw bytes start with `0x00` will decode to fewer bytes than the encoder produced. For example, encoding `b"\x00\x00\xab..."` (34 bytes) and decoding it back yields 32 bytes the two leading zeros are silently dropped.

## Reproduction

```python
from didkey import _b58decode

alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
payload = b"\x00\x00" + b"\xab" * 32
n = int.from_bytes(payload, "big")
encoded = ""
while n:
    n, rem = divmod(n, 58)
    encoded = alphabet[rem] + encoded
encoded = "11" + encoded  # two leading '1's for two 0x00 bytes

decoded = _b58decode(encoded)
assert len(decoded) == 34  # FAILS: returns 32 bytes, leading zeros lost
```

## Impact

Not currently reachable from real Ed25519 `did:key` identifiers the multicodec prefix `0xed01` never starts with `0x00`, so `public_key()` always receives 34-byte payloads that decode correctly. However:

- It violates the [base58btc spec](https://datatracker.ietf.org/doc/html/draft-msporny-base58-03#section-3) which requires leading `'1'` → `0x00` preservation
- Any future key type whose multicodec varint starts with a zero byte would silently fail the length check in `public_key()` and be rejected as malformed
- The test helper `_multibase` in `tests/_client.py` has the same bug on the encode side

## Suggested fix

```python
def _b58decode(raw: str) -> bytes:
    n = 0
    for ch in raw:
        digit = _B58_INDEX.get(ch)
        if digit is None:
            raise DidError(f"bad did:key: {ch!r} is not base58btc")
        n = n * 58 + digit
    leading = len(raw) - len(raw.lstrip("1"))
    payload = n.to_bytes((n.bit_length() + 7) // 8, "big") if n else b""
    return b"\x00" * leading + payload
```

All 369 existing tests pass with this fix.

### PR DESCRIPTION
**Pull request by evrendag**
## What

Fixes `_b58decode()` so leading base58btc `1` characters are restored as leading zero bytes instead of being lost during integer conversion.

Adds regression coverage for both a payload with multiple leading zero bytes and an all-zero base58btc value.

## Why

Base58btc defines leading `1` characters as leading `0x00` bytes. The previous integer-only decoder silently dropped those bytes, violating the encoding semantics even though current Ed25519 `did:key` values are not affected because their multicodec prefix is non-zero.

Closes #155.

## Scope

No protocol/API behavior changes for currently valid Ed25519 `did:key` identifiers; this makes the internal decoder spec-correct and future-safe.

### CODE REVIEW

File: src/didkey.py [hunk 1]
```diff
@@ -16,45 +16,19 @@
 import base64
 import re
```
Comment (alisemut0711-dev) [inline src/didkey.py:17]: Verified the fix against current `main` — `_b58decode` today returns `b'9'` for `"11z"` (leading `1` digits dropped by the integer round-trip), so the regression tests genuinely exercise the bug, and `public_key()` rejects every input the old decoder would have corrupted (demands 48 multibase chars starting `z`, 34-byte `\xed\x01`-prefixed decode), so the "no behavior change for valid DIDs" claim holds — I checked the boundary cases directly.

One suggestion to keep the diff focused: this hunk also removes the libsodium/backend rationale comment block (and the MULTIBASE_CHARS/SIG_CHARS notes), unrelated to the base58 decode fix. This repo treats comments as protocol rationale (AGENTS.md) and reviewers have pushed back on stripping them before. Keeping those deletions in a separate docs commit — or dropping them — would make the actual fix easier to review. The decode change + tests themselves are ready to go either way.

Comment (evrendag) [reply src/didkey.py:17]: Agreed. I restored the unrelated rationale/comments and current canonical signature constraint so the PR is focused on the base58btc decoder plus its regression tests. Thanks for checking the boundary cases.

**Comment by alisemut0711-dev** [comment]
Heads-up: it looks like a newer main has left this behind — GitHub shows `mergeable_state: dirty` (base: `7707cb...`, current main: `487caca`, which touches neither `didkey.py` nor the tests, so a straightforward `git rebase upstream/main` should clear it. If you'd like me to try a rebase from my side, happy to — otherwise no action needed from me.

**Comment by sv** [comment]
Closing as a duplicate of [#156](https://github.com/flop-labs/technocore-chat/pull/156), which fixes the same `_b58decode` leading-zero-byte loss (#155) in `src/didkey.py` and was filed four days earlier.

Worth flagging for whoever picks this up: [#444](https://github.com/flop-labs/technocore-chat/pull/444) also carries a base58 leading-zeroes fix, bundled with a CAS change for the room-claim races (#173/#155 in its title). Whichever lands first will conflict with the other, so they want reviewing together rather than in sequence.

#155 stays open.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by evrendag** [comment]
Thanks for the heads-up. Understood — #559 overlaps with #156, so closing this one as a duplicate is the right call. I’ll leave #155 open as requested and avoid creating a competing implementation while #156/#444 are being reviewed.
